### PR TITLE
CPM-554 create setSimpleReferenceEntityValue user intent

### PR DESCRIPTION
**Pull request by amigaire**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

- [x] Tests
- [ ] Delete EE branche once CE is merged

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductIntegration.php [hunk 1]
```diff
@@ -671,4 +748,28 @@ private function createAttributeOptions(string $attributeCode, string $optionCod
         }
         $this->get('pim_catalog.saver.attribute_option')->save($attributeOption);
     }
+
+    private function createReferenceEntity(string $referenceEntitycCode, array $labels = []): void
```
Comment (amigaire) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductIntegration.php:752]: check if can be used in clear reference_entity test ( l300~)

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductIntegration.php [hunk 2]
```diff
@@ -671,4 +748,28 @@ private function createAttributeOptions(string $attributeCode, string $optionCod
...
+        $createReferenceEntityCommand = new CreateReferenceEntityCommand($referenceEntitycCode, $labels);
+        $validator = $this->get('validator');
+        $violations = $validator->validate($createReferenceEntityCommand);
+        Assert::assertCount(0, $violations, \sprintf('The command is not valid: %s', $violations));
+        ($this->get('akeneo_referenceentity.application.reference_entity.create_reference_entity_handler'))(
+            $createReferenceEntityCommand
+        );
+    }
+
+    private function createRecords(string $referenceEntitycCode, array $recordCodes, array $labels = []): void
```
Comment (amigaire) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductIntegration.php:764]: remove $labels

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Specification/API/Command/UserIntent/SetSimpleReferenceEntityValueSpec.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+    function it_returns_the_channel_code()
+    {
+        $this->channelCode()->shouldReturn(null);
+    }
+
+    function it_returns_the_value()
+    {
+        $this->value()->shouldReturn('Akeneo');
+    }
```
Comment (amigaire) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Specification/API/Command/UserIntent/SetSimpleReferenceEntityValueSpec.php:46]: add test it_cannot_be_instantiated_with_an_empty_record_code

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductIntegration.php [hunk 4]
```diff
@@ -548,6 +531,81 @@ public function it_throws_an_exception_when_family_code_is_blank(): void
         $this->messageBus->dispatch($command);
     }
 
+    /** @test */
+    public function it_creates_a_product_with_an_existing_reference_entity()
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductIntegration.php:535]: In fact, you're creating a product with a record.
So it_successfully_creates_a_product_with_a_record should be better IMHO.

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductIntegration.php [hunk 5]
```diff
@@ -548,6 +531,81 @@ public function it_throws_an_exception_when_family_code_is_blank(): void
...
+        ]);
+        $this->messageBus->dispatch($command);
+
+        $this->clearDoctrineUoW();
+
+        $this->assertProductHasCorrectValueByAttributeCode('a_reference_entity_attribute', 'Akeneo');
+    }
+
+    /** @test */
+    public function it_updates_a_product_with_a_reference_entity()
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductIntegration.php:561]: it_successfully_sets_a_product_record_code

**Review by pierallard — APPROVED** [review]

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Specification/API/Command/UserIntent/SetSimpleReferenceEntityValueSpec.php [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        $this->localeCode()->shouldReturn(null);
+    }
+
+    function it_returns_the_channel_code()
+    {
+        $this->channelCode()->shouldReturn(null);
+    }
+
+    function it_returns_the_reord_code()
```
Comment (amigaire) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Specification/API/Command/UserIntent/SetSimpleReferenceEntityValueSpec.php:43]:
```suggestion
    function it_returns_the_record_code()
```

File: src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/SetSimpleReferenceEntityValueApplier.php [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    public function __construct(
+        private ObjectUpdaterInterface $productUpdater,
+    ) {
+    }
+
+    public function apply(UserIntent $userIntent, ProductInterface $product, int $userId): void
+    {
+        if (!$userIntent instanceof SetSimpleReferenceEntityValue) {
+            throw new \InvalidArgumentException('Not expected class');
+        }
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/SetSimpleReferenceEntityValueApplier.php:27]:
```suggestion
        Assert::isInstanceOf($userIntent, SetSimpleReferenceEntityValue::class);
```

**Review by mmetayer — APPROVED** [review]

**Pull request merged**